specs/chugsplash-config.md

```diff
@@ -0,0 +1,276 @@
+
+> Note that when using the ChugSplash Foundry library, users are limited to either Javascript or JSON.
+
+## Options
+- Project name
+- todo spec out other project options
+
+## Contracts
+JSON or Javascript object defining the contracts that should be deployed. Each key is a contracts reference name and its value is a contract definition.
+
+> Reference names are used internally by ChugSplash to identify specific instances of contracts. Each combination of a project name and contract reference name should be unique within the ChugSplash system.
+
+## Contract Definition
+JSON or Javascript object defining a contract to be deployed, it's constructor args, and variable definitions. Contract definitions must conform to the following format:
+
+```js
+ReferenceName: {
+  contract: 'LiteralContractName',
+  constructorArgs: {},
+  variables: {}
+}
+```
+
+## Contract Constructor Args
+A JSON or Javascript object defining a set of constructor args. All constructor arguements are required to be immutable variables...
+
+TODO finish defining constructor args spec
+
+## Contract Variables
+A JSON or Javascript object defining a set of ChugSplash input variables. ChugSplash is designed to remove the need for constructors or initializers by allowing the user to define the value of their variables directly during the deployment process. To ensure this process works reliably we perform input validation during the bundling process. Below we define the expected valid types. All other input types should be caught and rejected.
+
+Note that ChugSplash config files may be written in Javascript and therefore users are able to take advantage of various Javscript features and libraries to fetch values to be used in their ChugSplash config files.
+
+## uint & int
+For uints and ints, we allow two valid input types:
+- Base 10 string numbers: "1", "10", "1000000000000000000", "-100"
+- Base 10 literal numbers: 1, 10, 100, -100
+
+We enforce valid input ranges based on the specific byte size and sign of the target solidity variable. For example, if a variable is defined as uint8, we expect that values outside the range 0:255 to be caught and rejected during our bundling process. Likewise for int8, we expect that values outside the range -127:128 would be caught and rejected.
+
+Note that we explicitly do not support alternative base inputs such as hexadecimal for uint and int.
+
+### Examples
+```js
+// Solidity
+uint public x;
+
+// Javascript
+x: "1",
+```
+
+```js
+// Solidity
+uint public x;
+
+// Javascript
+x: 1,
+```
+
+```js
+// Solidity
+int public x;
+
+// Javascript
+x: -1,
+```
+
+```js
+// Solidity
+int public x;
+
+// Javascript
+x: "-1",
+```
+
+## Bytes32
```

Comment:
I think this should be "Fixed-size byte arrays" instead of "bytes32". In the Solidity docs: https://docs.soliditylang.org/en/latest/types.html